pkg/genericapiserver/storage_destinations.go

```diff
@@ -0,0 +1,189 @@
+	// of exposing one set of concepts.  autoscaling.HPA and extensions.HPA as a for instance
+	// The order of the slice matters!  It is the priority order of lookup for finding a storage location
+	CohabitatingResources map[unversioned.GroupResource][]unversioned.GroupResource
+
+	// APIResourceConfigSource indicates whether the *storage* is enabled, NOT the API
+	APIResourceConfigSource APIResourceConfigSource
+}
+
+const AllResources = "*"
+
+func NewStorageDestinations(defaultEtcdConfig etcdstorage.EtcdConfig, defaultSerializer runtime.NegotiatedSerializer, resourceEncodingConfig ResourceEncodingConfig, resourceEnablement APIResourceConfigSource) StorageDestinations {
+	return StorageDestinations{
+		DefaultEtcdConfig:       defaultEtcdConfig,
+		EtcdLocationOverrides:   map[unversioned.GroupResource][]string{},
+		EtcdPrefixOverrides:     map[unversioned.GroupResource]string{},
+		DefaultSerializer:       defaultSerializer,
+		SerializerOverrides:     map[unversioned.GroupResource]runtime.NegotiatedSerializer{},
+		ResourceEncodingConfig:  resourceEncodingConfig,
+		CohabitatingResources:   map[unversioned.GroupResource][]unversioned.GroupResource{},
+		APIResourceConfigSource: resourceEnablement,
+	}
+}
+
+func (s *StorageDestinations) SetEtcdLocation(groupResource unversioned.GroupResource, location []string) {
+	s.EtcdLocationOverrides[groupResource] = location
+}
+
+func (s *StorageDestinations) SetEtcdPrefix(groupResource unversioned.GroupResource, prefix string) {
+	s.EtcdPrefixOverrides[groupResource] = prefix
+}
+
+func (s *StorageDestinations) SetSerializer(groupResource unversioned.GroupResource, serializer runtime.NegotiatedSerializer) {
+	s.SerializerOverrides[groupResource] = serializer
+}
+
+// AddCohabitatingResources links resources together the order of the slice matters!  its the priority order of lookup for finding a storage location
+func (s *StorageDestinations) AddCohabitatingResources(groupResources ...unversioned.GroupResource) {
+	for _, resource := range groupResources {
+		s.CohabitatingResources[resource] = groupResources
+	}
+}
+
+func getAllResourcesAlias(resource unversioned.GroupResource) unversioned.GroupResource {
+	return unversioned.GroupResource{Group: resource.Group, Resource: AllResources}
+}
+
+func (s *StorageDestinations) getStorageGroupResource(groupResource unversioned.GroupResource) unversioned.GroupResource {
+	for _, potentialStorageResource := range s.CohabitatingResources[groupResource] {
+		if s.APIResourceConfigSource.AnyVersionOfResourceEnabled(potentialStorageResource) {
+			return potentialStorageResource
+		}
+	}
+
+	return groupResource
+}
+
+// Get finds the storage destination for the given group and resource. It will
+// return an error if the group has no storage destination configured.
+func (s *StorageDestinations) Get(groupResource unversioned.GroupResource) (storage.Interface, error) {
```

Comment:
New() is probably a better name?
With Get() I expect that I can call that method multiple times and I should get the same output.
This returns me a new instance every time, which is what I expect from New().